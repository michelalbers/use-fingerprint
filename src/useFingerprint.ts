import React from "react";
import Fingerprint2 from "fingerprintjs2";

export default function () {
  const [fingerprint, setFingerprint] = React.useState<string>(null);

  React.useEffect(() => {
    if (typeof window !== "object") return; // Run on the server side - ignore

    const setFingerprintFromComponents = () => {
      Fingerprint2.get((components) => {
        const fingerprint = Fingerprint2.x64hash128(
          components.map((c) => c.value).join(""),
          31
        );
        setFingerprint(fingerprint);
      });
    };

    // @ts-ignore
    if (window.requestIdleCallback) {
      // @ts-ignore
      window.requestIdleCallback(setFingerprintFromComponents);
    } else {
      setTimeout(setFingerprintFromComponents, 500);
    }
  }, []);

  return fingerprint;
}
