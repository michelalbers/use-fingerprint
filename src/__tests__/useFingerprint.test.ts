import Fingerprint2 from "fingerprintjs2";
import useFingerprint from "../useFingerprint";
import { request } from "http";
import { renderHook, act } from "@testing-library/react-hooks";

describe("useFingerprint hook", () => {
  beforeEach(() => {
    Fingerprint2.get = jest
      .fn()
      .mockImplementation((cb) => cb([{ key: "foo", value: "bar" }]));
  });

  it("should use requestIdleCallback when available", () => {
    spyOn(window, "setTimeout").and.callThrough();
    jest.useFakeTimers();

    // @ts-ignore
    window.requestIdleCallback = (cb) => cb();
    const { result } = renderHook(useFingerprint);

    act(() => {
      jest.runAllTimers();
    });

    expect(result.error).toBeUndefined();
    expect(typeof result.current).toEqual("string");
    expect(result.current).toMatchSnapshot();
    expect(window.setTimeout).not.toHaveBeenCalledWith(expect.anything(), 500);
  });

  it("should fallback to setTimeout if requestIdleCallback is not available", () => {
    spyOn(window, "setTimeout").and.callThrough();
    jest.useFakeTimers();

    // @ts-ignore
    window.requestIdleCallback = undefined;
    const { result } = renderHook(useFingerprint);

    act(() => {
      jest.runAllTimers();
    });

    expect(result.error).toBeUndefined();
    expect(typeof result.current).toEqual("string");
    expect(result.current).toMatchSnapshot();
    expect(window.setTimeout).toHaveBeenCalledWith(expect.anything(), 500);
  });
});
