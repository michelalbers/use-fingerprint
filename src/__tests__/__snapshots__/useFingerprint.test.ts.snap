// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFingerprint hook should fallback to setTimeout if requestIdleCallback is not available 1`] = `"14b8b56003cd9b5a48eaa8cbc909d853"`;

exports[`useFingerprint hook should use requestIdleCallback when available 1`] = `"14b8b56003cd9b5a48eaa8cbc909d853"`;
